conmon-rs/server/src/child_reaper.rs

```diff
@@ -132,28 +132,40 @@ impl ReapableChild {
         let tx_clone = exit_tx.clone();
 
         task::spawn_blocking(move || {
-            let wait_status = waitpid(Pid::from_raw(pid as pid_t), None);
-            match wait_status {
-                Ok(status) => {
-                    if let WaitStatus::Exited(_, exit_code) = status {
-                        debug!("Sending exit code to channel: {}", exit_code);
-                        if tx_clone.send(exit_code).is_err() {
-                            bail!("unable to send exit status")
+            const FAILED_EXIT_CODE: i32 = -3;
+            for i in 1..10 {
+                match waitpid(Pid::from_raw(pid as pid_t), None) {
+                    Ok(status) => {
+                        if let WaitStatus::Exited(_, exit_code) = status {
+                            debug!("Sending exit code to channel: {}", exit_code);
+                            tx_clone
+                                .send(exit_code)
+                                .map_err(|_| format_err!("unable to send exit status"))?;
+                            Self::write_to_exit_paths(exit_code, &exit_paths)
+                                .context("write to exit paths on successfull waitpid")?;
                         }
-                        Self::write_to_exit_paths(exit_code, &exit_paths)?;
+                        break;
                     }
-                }
-                Err(err) => {
-                    // TODO perhaps writing to the exit file anyway?
-                    // TODO maybe retry the waitpid?
-                    if err != Errno::ECHILD {
-                        bail!("caught error in reading for sigchld {}", err);
+                    Err(err) if err != Errno::ECHILD => {
+                        Self::write_to_exit_paths(FAILED_EXIT_CODE, &exit_paths)
+                            .context("write to exit paths on waitpid err != ECHILD")?;
+                        bail!("caught error in reading for sigchld {}", err)
                     }
-                    if tx_clone.send(-3).is_err() {
-                        bail!("unable to send exit status")
+                    Err(err) if err != Errno::EINTR => {
```

Comment:
wouldn't this fail on ECHILD now?

Reply:
Ooops this logic is incorrect, fixing it.